import glob


##### REFERENCE #####
REFDIR =        "/home/sunnymouse/Human_ref/"
FASTAREF =      REFDIR+"hg19_chrom/"
STARREFDIR =    REFDIR+"star/"
CHRNAME =       STARREFDIR+"chrName.txt"
GTFFILE =       REFDIR+"gencode.v19.annotation.gtf"

##### TOOLS #####

CONDA = "/home/sunnymouse/projects/PTES/python2.bio.yaml"
TOBIGBED = ""   # TODO add path to bedToBigBed

##### INPUT #####
INPUT_FOLDER = ""   # folder with BAM files
IDS, =         glob_wildcards(INPUT_FOLDER+"/{id}.bam")

PROJECT_FOLDER =  "/home/sunnymouse/projects/PTES/ENCODE"
DIRS =  ['fastq/','mapped/',]
FASTQ1S =        ['fastq/'+f+'.fq1' for id in IDS]
FASTQ2S =        ['fastq/'+f+'.fq2' for id in IDS]
MAPPED_SAM1 =     ['mapped/mate1_'+f+'.sam' for id in IDS]   # TODO add renaming for STAR _Aligned.out.sam
MAPPED_SAM2 =     ['mapped/mate2_'+f+'.sam' for id in IDS]
MAPPED_CHIM1 =    ['mapped/mate1_'+f+'.Chimeric.out.junction' for id in IDS]
MAPPED_CHIM2 =    ['mapped/mate2_'+f+'.Chimeric.out.junction' for id in IDS]


rule all:
	input: DIRS, FASTQ1S, FASTQ2S, MAPPED_SAM1, MAPPED_SAM2, MAPPED_CHIM1, MAPPED_CHIM2,

rule dirs:
	output: DIRS
	shell: "mkdir -p "+' '.join(DIRS)

##### BAM TO FASTQ #####
rule bam_to_fastq:
	input: bam="{folder}/{id}.bam".format(folder=INPUT_FOLDER), prefix="fastq/tmp_{id}",
	output: fq1="fastq/{id}.fq1", fq2="fastq/{id}.fq2",
	threads: 9
    conda: CONDA
	shell:
          """
	      samtools collate -uOn 256 {input.bam} {input.prefix} | samtools fastq -1 {output.fq1} -2 {output.fq2} -
	      """


##### ALIGNMENT #####
rule starindex:
	input: ref=FASTAREF, starref=STARREFDIR, gtf=GTFFILE,
	output: CHRNAME
    conda: CONDA
	shell: "STAR --limitGenomeGenerateRAM 32000000000 --runMode genomeGenerate --genomeDir {input.starref} --genomeFastaFiles {input.ref} --sjdbGTFfile {input.gtf} "

rule map:
	input:  fq1="fastq/{id}.fq1", fq2="fastq/{id}.fq2", starref=STARREFDIR, gtf=GTFFILE,
	output: "mapped/{sample}.sam"
	threads: 9
    conda: CONDA
	shell:
		"""
		STAR --genomeDir {input.starref} --outFileNamePrefix {wildcards.sample}_ --readFilesIn {input.sample} --runThreadN 24 --genomeLoad NoSharedMemory --outSAMattributes All --outSAMstrandField intronMotif --sjdbGTFfile {input.gtf}
		mv {wildcards.sample}_Aligned.out.sam {output}
		mv {wildcards.sample}_Log.final.out {wildcards.sample}_Log.out {wildcards.sample}_Log.progress.out {wildcards.sample}_SJ.out.tab starlogs
		"""
